presto-spi/src/main/java/com/facebook/presto/spi/block/StructBuilder.java

```diff
@@ -0,0 +1,226 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.Slices;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.unmodifiableList;
+
+public class StructBuilder
+{
+    private final BlockBuilderStatus builderStatus;
+    private final BlockBuilder builder;
+    private final List<Type> parameterTypes;
+    private final Iterator<Type> typeIterator;
+
+    public static StructBuilder arrayBuilder(Type elementType)
+    {
+        return new StructBuilder(asList(elementType));
+    }
+
+    public static StructBuilder mapBuilder(Type keyType, Type valueType)
+    {
+        return new StructBuilder(asList(keyType, valueType));
+    }
+
+    public static StructBuilder rowBuilder(List<Type> parameterTypes)
+    {
+        return new StructBuilder(parameterTypes);
+    }
+
+    public static StructBuilder builder(int expectedSize, List<Type> parameterTypes)
+    {
+        return new StructBuilder(expectedSize, parameterTypes);
+    }
+
+    private StructBuilder(List<Type> parameterTypes)
+    {
+        this(1024, parameterTypes);
+    }
+
+    private StructBuilder(int expectedSize, List<Type> parameterTypes)
+    {
+        this(new BlockBuilderStatus(expectedSize, expectedSize), parameterTypes);
+    }
+
+    private StructBuilder(BlockBuilderStatus builderStatus, List<Type> parameterTypes)
+    {
+        if (builderStatus == null) {
+            throw new NullPointerException("builderStatus is null");
+        }
+        if (parameterTypes == null) {
+            throw new NullPointerException("parameterTypes is null");
+        }
+        if (parameterTypes.isEmpty()) {
+            throw new IllegalArgumentException("parameterTypes is empty");
+        }
+        ArrayList<Type> types = new ArrayList<>();
+        for (Type type : parameterTypes) {
+            if (type == null) {
+                throw new NullPointerException("type element is null");
+            }
+            types.add(type);
+        }
+
+        this.builderStatus = builderStatus;
+        this.builder = new VariableWidthBlockBuilder(builderStatus);
+        this.parameterTypes = unmodifiableList(types);
+        this.typeIterator = new CycleIterator();
+    }
+
+    public static Block readBlock(SliceInput input)
+    {
+        return new VariableWidthBlockEncoding().readBlock(input);
+    }
+
+    public Slice build()
+    {
+        // VariableWidthBlockEncoding requires additional 4 bytes of element size and 1/8 of null bit per element
+        // So additional 5 bytes per element is large enough to serialize a whole block
+        int size = builderStatus.getSizeInBytes() + 5 * (builder.getPositionCount() + 1);
```

Comment:
BlockBuilderStatus.getSizeInBytes() is the current size in memory, which will be way bigger than necessary if there are only a few elements. Also, this logic is an implementation detail of VariableWithBlockBuilder. We should add a method to that class that returns a conservative estimate of the size required to serialize it.